### PR TITLE
New BG concept: fixed salinity, but time variable

### PR DESCRIPTION
**Pull request by jvollhueter**
Here is my first draft for the new belowground concept. Probably the two classes fixed salinity and fixed salinity t var still have to be combined, since many functions are contained in both. but what do you say to the general concept? i have uploaded an input file and an example setup for testing purposes.

### CODE REVIEW

File: TreeModelLib/BelowgroundCompetition/FixedSalinity_t_var/FixedSalinity_t_var.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    #  by porewater salinity. Only tree height aund leaf
+    #  water potential is needed\n
+    #  @param t_ini - initial time for next timestep \n
+    #  @param t_end - end time for next timestep
+    def prepareNextTimeStep(self, t_ini, t_end):
+        self._h_stem = []
+        self._r_crown = []
+        self._psi_leaf = []
+        self._xe = []
+        self.n += 1
```
[the remark this answers is not in the document]
Comment (jvollhueter) [reply TreeModelLib/BelowgroundCompetition/FixedSalinity_t_var/FixedSalinity_t_var.py:105]: Done

File: TreeModelLib/BelowgroundCompetition/FixedSalinity_t_var/FixedSalinity_t_var.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+    ## This function reads salinity from the control file.\n
+    def GetSalinity(self, args):
+        missing_tags = ["salinity_file", "type", "max_x", "min_x"]
+
+        for arg in args.iterdescendants():
+            tag = arg.tag
+            if tag == "salinity_file":
+                salinity_file = arg.text
+                # MARKER Fehlermeldung
```
[the remark this answers is not in the document]
Comment (jvollhueter) [reply TreeModelLib/BelowgroundCompetition/FixedSalinity_t_var/FixedSalinity_t_var.py:60]: Done

File: TreeModelLib/BelowgroundCompetition/FixedSalinity_t_var/FixedSalinity_t_var.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+"""
+@date: 2020-Today
+@author: ronny.peters@tu-dresden.de
+"""
+import numpy as np
+from TreeModelLib import TreeModel
```
[the remark this answers is not in the document]
Comment (jvollhueter) [reply TreeModelLib/BelowgroundCompetition/FixedSalinity_t_var/FixedSalinity_t_var.py:8]: Done

File: TreeModelLib/BelowgroundCompetition/FixedSalinity_t_var/FixedSalinity_t_var.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+
+class FixedSalinity_t_var(TreeModel):
```
[the remark this answers is not in the document]
Comment (jvollhueter) [reply TreeModelLib/BelowgroundCompetition/FixedSalinity_t_var/FixedSalinity_t_var.py:11]: Done

File: TreeModelLib/BelowgroundCompetition/FixedSalinity_t_var/FixedSalinity_t_var.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    #  @date: 2020 - Today
+    def __init__(self, args):
+        case = args.find("type").text
+        print("Initiate belowground competition of type " + case + ".")
+        self.GetSalinity(args)
+        print('________')
+        print('ini')
+        print('________')
+        if not hasattr(self, 'n'):
+            self.n = 0
```
[the remark this answers is not in the document]
Comment (jvollhueter) [reply TreeModelLib/BelowgroundCompetition/FixedSalinity_t_var/FixedSalinity_t_var.py:23]: I have to add the lines 22 + 23 to introduce the argument with time step number. Can I do this in that way:

```suggestion
    ## Fixed salinity in belowground competition concept.
    #  @param: Tags to define FixedSalinity: type, salinity
    #  @date: 2020 - Today
    def __init__(self, args):
        super().__ini__(args)
        if not hasattr(self, 'n'):
            self.n = 0
```

File: TreeModelLib/BelowgroundCompetition/FixedSalinity_t_var/FixedSalinity_t_var.py [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+    ## This function returns a list of salinity values for each tree,
+    # obtained by interpolation along a defined gradient which is defined
+    # for each time step
+    def getTreeSalinity(self):
+        self._salinity = self._salinity_over_t[self.n][1:]
+        return super().getTreeSalinity()
+
+    ## This function reads salinity from the control file.\n
+    def GetSalinity(self, args):
```
Comment (jvollhueter) [inline TreeModelLib/BelowgroundCompetition/FixedSalinity_t_var/FixedSalinity_t_var.py:28]: The function GetSalinity is quite similar to this one in parent class. Is it possible to use super somehow in this case? I think the problem is that the differences are in a for loop and a if statement...

Comment (jbathmann) [reply TreeModelLib/BelowgroundCompetition/FixedSalinity_t_var/FixedSalinity_t_var.py:28]: I would try to restructure the parent class. You could try to write everything depending on functions and just redefine the specific block.
Example Structure:

-Parent Class
 ```
class tolle_class(self):
  def tolle_function(self):
        self.a()
        self.b()
        self.c()
  
  def a(self):
      pass
  
  def b(self):
      pass
  
  def c(self):
      pass
```

- Child class
```

class tolles_kind(tolle_class):
    def b(self):
        print("Ich mache etwas anderes als meine eltern")
```

This way, "tolle_function" is different between parent and child

**Comment by jbathmann** [comment]
This PR was substituted by #180

**Pull request closed without merging**